Sources/NIOCore/AsyncAwaitSupport.swift

```diff
@@ -218,4 +218,77 @@ extension ChannelPipeline {
         try await self.addHandlers(handlers, position: position)
     }
 }
+
+public struct NIOTooManyBytesError: Error {
+     public init() {}
+ }
+
+@available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+extension AsyncSequence where Element: RandomAccessCollection, Element.Element == UInt8 {
+    /// Consumes an ``Swift/AsyncSequence`` of ``Swift/RandomAccessCollection``s into a single ``accumulationBuffer``.
+    /// - Parameters:
+    ///   - accumulationBuffer: buffer to write all the elements of `self` into
+    ///   - maxBytes: The maximum number of bytes this method is allowed to write into `accumulationBuffer`
+    /// - Throws: `NIOTooManyBytesError` if the the sequence contains more than `maxBytes`.
+    /// Note that previous elements of `self` might be already write to `accumulationBuffer`.
+    @inlinable
+    public func collect(
+        into accumulationBuffer: inout ByteBuffer,
+        maxBytes: Int
+    ) async throws {
+        var bytesRead = 0
+        for try await fragment in self {
+            let fragmentSize = fragment.count
+            bytesRead += fragmentSize
+            guard bytesRead <= maxBytes else {
+                throw NIOTooManyBytesError()
+            }
+            accumulationBuffer.writeBytes(fragment)
+        }
+    }
+    
+    /// Consumes an ``Swift/AsyncSequence`` of ``Swift/RandomAccessCollection``s into a single ``NIO/ByteBuffer``.
+    /// - Parameters:
+    ///   - maxBytes: The maximum number of bytes this method is allowed to write into `accumulationBuffer`
+    /// - Throws: `NIOTooManyBytesError` if the the sequence contains more than `maxBytes`.
+    @inlinable
+    public func collect(
+        maxBytes: Int
+    ) async throws -> ByteBuffer {
+        var accumulationBuffer = ByteBuffer()
```

Comment:
I think, we should allocate some space here. Otherwise we realloc a lot in the beginning.

Reply:
But how much? 
For the case where `AsyncSequence.Element == ByteBuffer` I have just added an optimisation to use the first element as the `accumulationBuffer`. Do you think this would be enough?

Reply:
Historically we have always suggested that users use an allocator provided by the channel to create buffers. Of course we now have a handful of `init`s for creating `ByteBuffer`s without an allocator which is certainly useful for testing etc. 

Is it our position now that we're happy for folks (and indeed us) to create `ByteBuffer`s without an allocator?

Reply:
I'm inclined to keep this API, but to have it take a `ByteBufferAllocator`. That way we preserve the implication that we may one day do something sensible with this type.

Reply:
And I do agree with `@fabianfett` that we should pre-allocate some space here. Maybe 512 bytes? Just to skip a few doublings.

Reply:
`512` sounds good to me. I would also be okay with `1024`. My feeling is generally that swift is so memory efficient, that we can gladly trade a little more memory consumption for better performance.

Reply:
increased to `1024`